src/Behat/Testwork/Argument/MixedArgumentOrganiser.php

```diff
@@ -217,29 +216,59 @@ private function prepareTypehintedArguments(array $parameters, array $typehinted
      * @param  ReflectionParameter[] $parameters Constructor Arguments
      * @return ReflectionParameter[]             Filtered $parameters
      */
-    private function filterApplicableTypehintedParameters(array $parameters)
+    private function filterApplicableTypehintedParameters(array $parameters) : array
     {
-        $filtered = array();
+        return array_filter($parameters,
+            function($parameter, $num) {
+                return !$this->isArgumentDefined($num)
+                && $this->getReflectionClassesFromParameter($parameter);
+            },
+            ARRAY_FILTER_USE_BOTH
+        );
 
-        foreach ($parameters as $num => $parameter) {
-            if ($this->isArgumentDefined($num)) {
-                continue;
+    }
+
+    /**
+     * @return ReflectionClass[]
+     */
+    private function getReflectionClassesFromParameter(\ReflectionParameter $parameter): array
+    {
+        $classes = [];
+
+        if (!$parameter->hasType()) {
+            return $classes;
+        }
+
+        $type = $parameter->getType();
+
+        if ($type instanceof \ReflectionNamedType) {
+            $types = [$type];
+        }
+        elseif ($parameter->getType() instanceof \ReflectionUnionType) {
+            $types = $type->getTypes();
```

Comment:
are we sure that this always returns a `\ReflectionNamedType[]` ?

Reply:
ReflectionUnionType is only documented in https://wiki.php.net/rfc/union_types_v2 so far where it's documented as returning `ReflectionType[]` It seems to always be ReflectionNamedType currently as we don't have unions of unions.

It won't hurt to add a check here as it could break in future with generics etc.